unbounded PCollections writes to files support in iobase derived IOs:  AvroIO, ParquetIO , TextIO, TFRecordIO

razvanculea opened the pull request:
iobase patch for writing unbounded PColl sources.
Implement unbounded PCollections writes to files support in iobase derived IOs:  AvroIO, ParquetIO , TextIO, TFRecordIO
Tests for streming PCollections writes to files in AvroIO, ParquetIO , TextIO, TFRecordIO

------------------------

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[![Build python source distribution and wheels](https://github.com/apache/beam/actions/workflows/build_wheels.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[![Python tests](https://github.com/apache/beam/actions/workflows/python_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[![Java tests](https://github.com/apache/beam/actions/workflows/java_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)
[![Go tests](https://github.com/apache/beam/actions/workflows/go_tests.yml/badge.svg?event=schedule&&?branch=master)](https://github.com/apache/beam/actions?query=workflow%3A%22Go+tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI or the [workflows README](https://github.com/apache/beam/blob/master/.github/workflows/README.md) to see a list of phrases to trigger workflows.

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1235,10 +1430,20 @@ def _finalize_write(
   outputs = sink.finalize_write(
-      init_result, write_results + extra_shards, pre_finalize_results)
+      init_result, write_results + extra_shards, pre_finalize_results, w)
+  outputs = list(outputs)
+  _LOGGER.info("*** _finalize_write outputs %s",outputs)
```
The returned generator should not be materialized into a list afaik.
```suggestion
```

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1235,10 +1430,20 @@ def _finalize_write(
   if outputs:
-    return (
-        window.TimestampedValue(v, timestamp.MAX_TIMESTAMP) for v in outputs)
+    if not isinstance(w, window.GlobalWindow):
+      #handle windowed finalize
+      yield (
+          window.TimestampedValue(v, 
+              timestamp=self.window.start,
+              windows=[w]
+            ) for v in outputs)
+    else:
+      return (
+          window.TimestampedValue(v, timestamp.MAX_TIMESTAMP) for v in outputs)
```
The constructor of `TimestampedValue` has no argument `windows` afaik and you likely want to shift these resulting elements to the window's end time.
```suggestion
      return (
          window.TimestampedValue(v, w.end()) for v in outputs)
```

razvanculea replied:
w.end, but it's true that the result values should be at the end of the window

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
+        )
+        finalized_write_result_coll = (
+          pre_finalized_write_result_coll 
+          #| 'LogElements pre_finalized_write_result_coll' >> LogElements(prefix="pre_finalized_write_result_coll :",with_window=True,with_timestamp=True,level=logging.INFO)
```
Remove
```suggestion
```

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
+            | 'WriteWindowedBundles' >> core.ParDo(
+                _WriteWindowedBundleDoFn(sink=self.sink,per_key=True), 
+                AsSingleton(init_result_window_coll))
+            #| 'LogElements' >> LogElements(prefix="after WriteWindowedBundles :",with_window=True,with_timestamp=True,level=logging.INFO)
```
Remove
```suggestion
```

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
+        write_result_coll = (
+          keyed_pcoll
+            | 'Group by random key' >> core.GroupByKey()
+            #| 'LogElements before WriteWindowedBundles' >> LogElements(prefix="before WriteWindowedBundles :",with_window=True,with_timestamp=True,level=logging.INFO)
```
Remove
```suggestion
```

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
+            | 'Pair init gbk' >> core.GroupByKey()
+            | 'InitializeWindowedWrite' >> core.Map(
+              lambda _, sink: sink.initialize_write(), self.sink)
+            #| 'LogElements init_result_window_coll' >> LogElements(prefix="init_result_window_coll :",with_window=True,with_timestamp=True,level=logging.INFO)
```
Remove
```suggestion
```

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
           keyed_pcoll
-          | core.WindowInto(window.GlobalWindows())
-          | core.GroupByKey()
-          | 'WriteBundles' >> core.ParDo(
-              _WriteKeyedBundleDoFn(self.sink), AsSingleton(init_result_coll)))
+            | 'Pair init' >> core.Map(lambda x: (None, x))
+            | 'Pair init gbk' >> core.GroupByKey()
+            | 'InitializeWindowedWrite' >> core.Map(
+              lambda _, sink: sink.initialize_write(), self.sink)
```
This doesn't have to depend on the main input, right?
I'm almost certain a globally windowed side input should suffice.

sjvanrossum commented on sdks/python/apache_beam/io/iobase.py:
```diff
@@ -1127,47 +1130,155 @@ def __init__(self, sink: Sink) -> None:
+            | 'WriteBundles' >> core.ParDo(
+                _WriteKeyedBundleDoFn(self.sink), AsSingleton(init_result_coll)))
+      else: #unbounded PCollection needes to be written per window
+        if isinstance(pcoll.windowing.windowfn, window.GlobalWindows):
+          widowed_pcoll = (
+            pcoll
+              | core.WindowInto(window.FixedWindows(self.sink.triggering_frequency),
+                                trigger=beam.transforms.trigger.AfterWatermark(),
+                                accumulation_mode=beam.transforms.trigger.AccumulationMode.DISCARDING,
+                                allowed_lateness=beam.utils.timestamp.Duration(seconds=0))
+          )
```
If the input is globally windowed a trigger within the global window set to repeat every `sink.triggering_frequency` would be what I'd expect instead of windowing into `FixedWindows`.

razvanculea replied:
that would be an alternative way to implement. 
The windowing will help have elements of the same window in the same file, vs group into batches & trigger will have elements as they come (possibly mixed)

shunping commented:
This PR is related to #34814 and may also fix #25598.

liferoad commented:
is this PR tested with DataflowRunner?

razvanculea commented:
> is this PR tested with DataflowRunner?

@liferoad i've added the iobase_it_test from PR 34814 and it runs on Dataflow (it's actually using TextIO for write), but i don't have an IT for TextIO/parquetIO/AvroIO/TFRecordIO, I can try to add them, but i can't use TestStream in Dataflow, so i'll need to use PubSub public topics , right ?

liferoad commented:
> > is this PR tested with DataflowRunner?
> 
> @liferoad i've added the iobase_it_test from PR 34814 and it runs on Dataflow (it's actually using TextIO for write), but i don't have an IT for TextIO/parquetIO/AvroIO/TFRecordIO, I can try to add them, but i can't use TestStream in Dataflow, so i'll need to use PubSub public topics , right ?



> > is this PR tested with DataflowRunner?
> 
> @liferoad i've added the iobase_it_test from PR 34814 and it runs on Dataflow (it's actually using TextIO for write), but i don't have an IT for TextIO/parquetIO/AvroIO/TFRecordIO, I can try to add them, but i can't use TestStream in Dataflow, so i'll need to use PubSub public topics , right ?

I forgot whether periodic impulse can trigger https://github.com/apache/beam/issues/25598. You could try that. And better add some checks about the output results.

razvanculea commented:
> > > is this PR tested with DataflowRunner?
> > 
> > 
> > @liferoad i've added the iobase_it_test from PR 34814 and it runs on Dataflow (it's actually using TextIO for write), but i don't have an IT for TextIO/parquetIO/AvroIO/TFRecordIO, I can try to add them, but i can't use TestStream in Dataflow, so i'll need to use PubSub public topics , right ?
> 
> > > is this PR tested with DataflowRunner?
> > 
> > 
> > @liferoad i've added the iobase_it_test from PR 34814 and it runs on Dataflow (it's actually using TextIO for write), but i don't have an IT for TextIO/parquetIO/AvroIO/TFRecordIO, I can try to add them, but i can't use TestStream in Dataflow, so i'll need to use PubSub public topics , right ?
> 
> I forgot whether periodic impulse can trigger #25598. You could try that. And better add some checks about the output results.

@liferoad, I added an IT in parquetio_it_test.py using PeriodicImpulse and it's working fine on Dataflow

Did not add IT for AvroIO/TFRecordIO because there is no IT at all. I could do like for parquetIO later, if we deem this really a plus.

razvanculea commented:
Improved support for user windowing.
If an unbounded PCollection with GlobalWindow is written:
* without setting triggering_frequency => throw an exception
* if triggering_frequency is set, then the output will be written per requested frequency

If an unbounded PCollection with a user set windowing is written:
* without setting triggering_frequency => tthen the output will be written per user defined window
* if triggering_frequency is set, then the output will be written per requested frequency

If num_shards is set : per triggering_frequency/user window, the output is written in num_shards
If num_shards is not set : per triggering_frequency/user window, the output is written with 1 file per bundle

shunping commented:
@razvanculea, sorry about the delay of the review. Do you mind merging to the latest master so that the tests are re-triggered.

shunping commented:
waiting on author

razvanculea commented:
@shunping don't review this anymore. 
I'll break it into IObase+FileBasedSync PR, and 1 PR per IO (Parquet,Avro,TFRecord,Text)
As @sjvanrossum pointed out I could avoid breaking changes, in case there are any private developped IOs that currently work for batch. These would not work for streaming without changes, but would still work as before on batch with newer SDK.

shunping commented:
sg! Feel free to tag me when they are ready.
